### DISCUSSION TITLE
4.1.0

### DISCUSSION THREAD

**Discussion by mklemm** [mklemm/jaxb2-rich-contract-plugin#75] [General]
Fixed bugs #46  and #68  as well as some other regressions and instabilities introduced in the hurry to get out 4.0.0, because JAXB 4 compliance was to priority.
Added more test cases and improved @at055612 s test runner class to regularly run for test runs.
Some minor code cleanup as well.

**Full Changelog**: https://github.com/mklemm/jaxb2-rich-contract-plugin/compare/jaxb2-rich-contract-plugin-4.0.0...4.1.0

<hr /><em>This discussion was created from the release <a href='https://github.com/mklemm/jaxb2-rich-contract-plugin/releases/tag/4.1.0'>4.1.0</a>.</em>

**Comment by at055612** [comment]
Thanks for this.  I have been hoping for migration from `javax` to `jakarta`.

>Added more test cases and improved @at055612 s test runner class to regularly run for test runs.

As far as I can see this only checks that the generated code compiles, whereas `ChoiceTest` and `ReservedWordTest` ran tests against the generated code to verify the objects created by the builders were correct.

---

The changes from 2.1.0 => 4.1.0 seem to have broken choice elements with a common supertype.
Run `testGenerateAll` then see `JavadocAnnotations.ChoiceOfElementsTwo.Builder` in
`target/test/generated-sources/xjc/all/com/kscs/jaxb2/contract/test/JavadocAnnotations.java`.

The Builder now has these fields
```java
        public static class Builder<_B >implements Buildable
        {
            ...
            private Transport transport;
            private Bike.Builder<JavadocAnnotations.ChoiceOfElementsTwo.Builder<_B>> bike;
            private Car.Builder<JavadocAnnotations.ChoiceOfElementsTwo.Builder<_B>> car;
```
whereas before I think it had only
```java
            private Buildable transport;
```

If you call `withTransport(aBike)` it works.  If you call `withBike(aBike)` then `aBike` is lost when `.build()` is called on `ChoiceOfElementsTwo`.

See the following methods to see that Bike and Car are being lost.
```java
            protected<_P extends JavadocAnnotations.ChoiceOfElementsTwo >_P init(final _P _product) {
                _product.transport = this.transport;
                return _product;
            }
            public JavadocAnnotations.ChoiceOfElementsTwo.Builder<_B> withTransport(final Transport transport) {
                this.transport = transport;
                return this;
            }
            public JavadocAnnotations.ChoiceOfElementsTwo.Builder<_B> withBike(final Bike bike) {
                this.bike = ((bike == null)?null:new Bike.Builder<>(this, bike, false));
                return this;
            }
            public JavadocAnnotations.ChoiceOfElementsTwo.Builder<_B> withCar(final Car car) {
                this.car = ((car == null)?null:new Car.Builder<>(this, car, false));
                return this;
            }
            @Override
            public JavadocAnnotations.ChoiceOfElementsTwo build() {
                if (_storedValue == null) {
                    return this.init(new JavadocAnnotations.ChoiceOfElementsTwo());
                } else {
                    return ((JavadocAnnotations.ChoiceOfElementsTwo) _storedValue);
                }
            }
```
`this.bike` and `this.car` are being set by the `with` methods but never being read.

**Reply by at055612** [reply]
It was possibly broken in this commit 3623521

**Reply by mklemm** [reply]
This is entirely possible, I removed some of the new choice code because it caused uncompilable code to be generated in some scenarios that we used the plugin for. Will revisit this ASAP. Priority recently was to change the group-interface generating strategy which caused trouble for a lot of people.

**Reply by at055612** [reply]
I understand.  Your plugin has to handle a large number of different use cases within the xsd so when I originally made the change for choice I feared it may break other cases that I wasn't aware of.  I'm sorry if this was the case.

Given the many cases that have to be supported I think the repo would benefit from a test framework that allows for junits to be written against the code generated by a given xsd/xjb pair.

**Reply by mklemm** [reply]
Yes, there is the jaxb2-rich-contract-plugin-test repository, but I'm not allowed to push our internal XSDs there, so there's a gap...

**Reply by at055612** [reply]
I can't see the repo `jaxb2-rich-contract-plugin-test` on github.  Is it public?

**Reply by mklemm** [reply]
Oh sorry, you're right, it's private... I'll have to have a look whether it contains undisclosable information, then I'll open it.

**Comment by mklemm** [comment]
Hi, I now changed the repo structure so that a test module is inlcuded.

**Reply by at055612** [reply]
Thanks, I'll take a look.

**Comment by at055612** [comment]
> I removed some of the new choice code because it caused uncompilable code to be generated in some scenarios that we used the plugin for. 

Would you be able to add in an xsd that demonstrates the scenarios that were broken by my choice change?  Then I may be able to fix the plugin so it works for both my case and yours.

I am really keen to use the latest version of your lib with `jakarta` packages but can't until the choice issue is resolved.

**Reply by mklemm** [reply]
Did you try the new version already? I put the XSD from your tests in the standard build process and it now seems to work ok...
Unfortunately, because the XSDs where I found the problems originated from a customer's project, I am not allowed to publish them here. As soon as I get the time, I'll create a synthetic example ...

**Reply by at055612** [reply]
I hadn't done, but have now.  The test module is much better :+1: 

The problem still remains though unfortunately.  I have raised an issue explaining the problem.  https://github.com/mklemm/jaxb-rich-contract-plugin/issues/78
